#include "weather_forecast.h"
#include <QMessageBox>
#include <QDebug>
#include <QJsonDocument>
#include <QJsonArray>
#include <QJsonObject>

weather_forecast::weather_forecast(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::weather_forecastClass())
{
    ui->setupUi(this);
    setWindowIcon(QIcon(":/weather_forecast/icon/1_1.png"));
    geoManager = new QNetworkAccessManager(this);
    weatherManager = new QNetworkAccessManager(this);

    // å½“ QNetworkAccessManager å®Œæˆä¸€ä¸ªç½‘ç»œè¯·æ±‚æ—¶ï¼Œä¼šå‘å‡º finished(QNetworkReply*) ä¿¡å·
    connect(geoManager, &QNetworkAccessManager::finished, this, &weather_forecast::onGeoReplyFinished);
    connect(weatherManager, &QNetworkAccessManager::finished, this, &weather_forecast::onWeatherReplyFinished);
    // è¾“å…¥æ¡†å›è½¦æ—¶è§¦å‘æŸ¥è¯¢
    connect(ui->cityInput, &QLineEdit::returnPressed, this, &weather_forecast::on_searchButton_clicked);
}

weather_forecast::~weather_forecast()
{
    delete ui;
}

// geoManager å‘é€ç½‘ç»œè¯·æ±‚
void weather_forecast::on_searchButton_clicked() {
    QString city = ui->cityInput->text().trimmed();
    if (city.isEmpty()) {
        QMessageBox::warning(this, "é”™è¯¯", "è¯·è¾“å…¥åŸå¸‚åï¼");
        return;
    }

    ui->currentWeatherLabel->setText("æ­£åœ¨æŸ¥è¯¢ä½ç½®...");
    ui->forecastList->clear();

    // è·å–ç»çº¬åº¦ æ˜¯é€šè¿‡open weather map API è·å–åœ°ç†æ•°æ®
    // åœ¨é€šè¿‡åŸå¸‚åè·å–åœ°ç†æ•°æ®,ä¹‹ååœ¨ç”¨å…è´¹çš„å¤©æ°”apiè·å–å¤©æ°”æ•°æ®
    QString geoUrl = QString("http://api.openweathermap.org/geo/1.0/direct?q=%1&limit=1&appid=%2")
        .arg(city, apiKey);

    // å‘é€ç½‘ç»œè¯·æ±‚ è·å–ç»çº¬åº¦ jsonæ•°æ®
    geoManager->get(QNetworkRequest(QUrl(geoUrl)));
}

// geoManager è·å–ç»çº¬åº¦æˆåŠŸåï¼Œè§¦å‘weatherManagerè·å–å¤©æ°”æ•°æ®
void weather_forecast::onGeoReplyFinished(QNetworkReply* reply)
{
    if (reply->error() != QNetworkReply::NoError) {
        QMessageBox::critical(this, "é”™è¯¯", "ä½ç½®æŸ¥è¯¢å¤±è´¥: " + reply->errorString());
        reply->deleteLater();
        return;
    }

    QByteArray data = reply->readAll();
    QJsonDocument doc = QJsonDocument::fromJson(data);
    QJsonArray array = doc.array();

    if (array.isEmpty()) {
        QMessageBox::warning(this, "é”™è¯¯", "æœªæ‰¾åˆ°è¯¥åŸå¸‚ï¼");
        reply->deleteLater();
        return;
    }

    QJsonObject obj = array.first().toObject();
    double lat = obj["lat"].toDouble();     // çº¬åº¦
    double lon = obj["lon"].toDouble();     // ç»åº¦
    cityName = obj["name"].toString() + ", " + obj["country"].toString();

    reply->deleteLater();

    QString weatherUrl = QString("https://api.open-meteo.com/v1/forecast?"
        "latitude=%1&longitude=%2"
        "&current=temperature_2m,weather_code"
        "&daily=temperature_2m_max,temperature_2m_min,weather_code"
        "&timezone=auto")
        .arg(lat).arg(lon);

    weatherManager->get(QNetworkRequest(QUrl(weatherUrl)));
    ui->currentWeatherLabel->setText(QString("æ­£åœ¨åŠ è½½ %1 çš„å¤©æ°”...").arg(cityName));
}

// weatherManager è·å–å¤©æ°”æˆåŠŸåï¼Œè§¦å‘uiæ›´æ–°
void weather_forecast::onWeatherReplyFinished(QNetworkReply* reply)
{
    if (reply->error() != QNetworkReply::NoError) {
        QMessageBox::critical(this, "é”™è¯¯", "å¤©æ°”æŸ¥è¯¢å¤±è´¥: " + reply->errorString());
        reply->deleteLater();
        return;
    }

    QByteArray data = reply->readAll();
    QJsonDocument doc = QJsonDocument::fromJson(data);
    QJsonObject root = doc.object();

    qDebug() << data; 
    //ui->textEditdebug->setText(data);

    // å½“å‰å¤©æ°”
    QJsonObject current = root["current"].toObject();
    double temp = current["temperature_2m"].toDouble();
    int code = current["weather_code"].toInt();
    ui->currentWeatherLabel->setText(cityName + QString("å½“å‰å¤©æ°”ï¼š%1Â°Cï¼Œ%2")
        .arg(temp, 0, 'f', 1)
        .arg(weatherDescription(code)));

    // é¢„æŠ¥
    ui->forecastList->clear();
    QJsonObject daily = root["daily"].toObject();
    QJsonArray dates = daily["time"].toArray();
    QJsonArray maxT = daily["temperature_2m_max"].toArray();
    QJsonArray minT = daily["temperature_2m_min"].toArray();
    QJsonArray codes = daily["weather_code"].toArray();

    for (int i = 0; i < dates.size(); ++i) {
        QString date = dates[i].toString();
        double max = maxT[i].toDouble();
        double min = minT[i].toDouble();
        int c = codes[i].toInt();
        ui->forecastList->addItem(QString("%1: æœ€é«˜ %2Â°Cï¼Œæœ€ä½ %3Â°Cï¼Œ%4")
            .arg(date).arg(max, 0, 'f', 1).arg(min, 0, 'f', 1).arg(weatherDescription(c)));
    }

    reply->deleteLater();
}

QString weather_forecast::weatherDescription(int code)
{
    switch (code) {
    case 0:  return "â˜€ï¸ æ™´æœ—";  // Clear weather
    case 1:
    case 2:  return "â›… éƒ¨åˆ†å¤šäº‘";  // Partly cloudy
    case 3:  return "â˜ï¸ é˜´å¤©";  // Overcast
    case 4:  return "ğŸŒ«ï¸ çƒŸé›¾";  // Visibility reduced by smoke
    case 5:  return "ğŸŒ«ï¸ éœ¾";  // Haze
    case 6:  return "ğŸ’¨ æµ®å°˜";  // Widespread dust in suspension
    case 7:  return "ğŸ’¨ æ‰¬æ²™";  // Dust or sand raised by wind
    case 8:  return "ğŸŒªï¸ å°˜å·é£";  // Well developed dust whirl(s)
    case 9:  return "ğŸŒªï¸ æ²™å°˜æš´";  // Duststorm or sandstorm
    case 10: return "ğŸŒ«ï¸ è–„é›¾";  // Mist
    case 11:
    case 12: return "ğŸŒ«ï¸ æµ…é›¾";  // Patches shallow fog / More or less continuous shallow fog
    case 13: return "âš¡ é—ªç”µ";  // Lightning visible, no thunder heard
    case 14:
    case 15: return "ğŸŒ§ï¸ è¿œé™æ°´";  // Precipitation within sight, not reaching ground / reaching ground, distant
    case 16: return "ğŸŒ§ï¸ è¿‘é™æ°´";  // Precipitation within sight, reaching ground, near
    case 17: return "âš¡ é›·æš´æ— é™æ°´";  // Thunderstorm, no precipitation
    case 18: return "ğŸŒªï¸ é£‘";  // Squalls
    case 19: return "ğŸŒªï¸ æ¼æ–—äº‘";  // Funnel cloud(s)
    case 20: return "ğŸŒ¦ï¸ æ¯›æ¯›é›¨";  // Drizzle or snow grains not shower
    case 21: return "ğŸŒ§ï¸ é›¨";  // Rain not freezing
    case 22: return "â„ï¸ é›ª";  // Snow
    case 23: return "ğŸŒ¨ï¸ é›¨å¤¹é›ª";  // Rain and snow or ice pellets
    case 24: return "â„ï¸ å†»é›¨";  // Freezing drizzle or freezing rain
    case 25: return "ğŸŒ¦ï¸ é˜µé›¨";  // Shower(s) of rain
    case 26: return "â„ï¸ é˜µé›ª";  // Shower(s) of snow, or of rain and snow
    case 27: return "âšª å†°é›¹";  // Shower(s) of hail
    case 28: return "ğŸŒ«ï¸ é›¾";  // Fog or ice fog
    case 29: return "âš¡ é›·æš´";  // Thunderstorm
    case 30: return "ğŸŒªï¸ æ²™å°˜æš´å‡å¼±";  // Slight or moderate duststorm, decreasing
    case 31: return "ğŸŒªï¸ æ²™å°˜æš´";  // Slight or moderate duststorm, no change
    case 32: return "ğŸŒªï¸ æ²™å°˜æš´å¢å¼º";  // Slight or moderate duststorm, increasing
    case 33: return "ğŸŒªï¸ å¼ºæ²™å°˜æš´å‡å¼±";  // Severe duststorm, decreasing
    case 34: return "ğŸŒªï¸ å¼ºæ²™å°˜æš´";  // Severe duststorm, no change
    case 35: return "ğŸŒªï¸ å¼ºæ²™å°˜æš´å¢å¼º";  // Severe duststorm, increasing
    case 36:
    case 38: return "â„ï¸ ä½å¹é›ª";  // Slight or moderate blowing snow, low / high
    case 37:
    case 39: return "â„ï¸ å¼ºå¹é›ª";  // Heavy drifting snow
    case 40: return "ğŸŒ«ï¸ è¿œé›¾";  // Fog at a distance
    case 41: return "ğŸŒ«ï¸ ç‰‡çŠ¶é›¾";  // Fog in patches
    case 42:
    case 44:
    case 46: return "ğŸŒ«ï¸ é›¾";  // Fog, sky visible
    case 43:
    case 45:
    case 47: return "ğŸŒ«ï¸ æµ“é›¾";  // Fog, sky invisible
    case 48: return "â„ï¸ é›¾å‡‡";  // Fog, depositing rime, sky visible
    case 49: return "â„ï¸ æµ“é›¾å‡‡";  // Fog, depositing rime, sky invisible
    case 50:
    case 51: return "ğŸŒ¦ï¸ é—´æ­‡æ€§å°æ¯›æ¯›é›¨";  // Drizzle, not freezing, intermittent slight / continuous
    case 52:
    case 53: return "ğŸŒ¦ï¸ æŒç»­æ€§æ¯›æ¯›é›¨";  // Drizzle, not freezing, intermittent moderate / continuous
    case 54:
    case 55: return "ğŸŒ§ï¸ é—´æ­‡æ€§å¤§æ¯›æ¯›é›¨";  // Drizzle, not freezing, intermittent heavy / continuous
    case 56: return "â„ï¸ å†»æ¯›æ¯›é›¨";  // Drizzle, freezing, slight
    case 57: return "â„ï¸ ä¸­åˆ°å¼ºå†»æ¯›æ¯›é›¨";  // Drizzle, freezing, moderate or heavy
    case 58: return "ğŸŒ¦ï¸ æ¯›æ¯›é›¨å’Œå°é›¨";  // Drizzle and rain, slight
    case 59: return "ğŸŒ§ï¸ æ¯›æ¯›é›¨å’Œä¸­åˆ°å¤§é›¨";  // Drizzle and rain, moderate or heavy
    case 60:
    case 61: return "ğŸŒ¦ï¸ é—´æ­‡æ€§å°é›¨";  // Rain, not freezing, intermittent slight / continuous
    case 62:
    case 63: return "ğŸŒ§ï¸ é—´æ­‡æ€§ä¸­é›¨";  // Rain, not freezing, intermittent moderate / continuous
    case 64:
    case 65: return "ğŸŒ§ï¸ é—´æ­‡æ€§å¤§é›¨";  // Rain, not freezing, intermittent heavy / continuous
    case 66: return "â„ï¸ å†»é›¨";  // Rain, freezing, slight
    case 67: return "â„ï¸ ä¸­åˆ°å¼ºå†»é›¨";  // Rain, freezing, moderate or heavy
    case 68: return "ğŸŒ¨ï¸ é›¨å¤¹é›ª";  // Rain or drizzle and snow, slight
    case 69: return "ğŸŒ¨ï¸ ä¸­åˆ°å¤§é›¨å¤¹é›ª";  // Rain or drizzle and snow, moderate or heavy
    case 70:
    case 71: return "â„ï¸ é—´æ­‡æ€§å°é›ª";  // Snowflakes, intermittent slight / continuous
    case 72:
    case 73: return "â„ï¸ é—´æ­‡æ€§ä¸­é›ª";  // Snowflakes, intermittent moderate / continuous
    case 74:
    case 75: return "â„ï¸ é—´æ­‡æ€§å¤§é›ª";  // Snowflakes, intermittent heavy / continuous
    case 76: return "â„ï¸ å†°æ™¶";  // Diamond dust
    case 77: return "â„ï¸ é›ªç²’";  // Snow grains
    case 78: return "â„ï¸ æ˜ŸçŠ¶é›ªæ™¶";  // Isolated star-like snow crystals
    case 79: return "â„ï¸ å†°ç²’";  // Ice pellets
    case 80: return "ğŸŒ¦ï¸ å°é˜µé›¨";  // Rain shower(s), slight
    case 81: return "ğŸŒ§ï¸ ä¸­åˆ°å¤§é˜µé›¨";  // Rain shower(s), moderate or heavy
    case 82: return "ğŸŒ§ï¸ å¼ºé˜µé›¨";  // Rain shower(s), violent
    case 83: return "ğŸŒ¨ï¸ å°é˜µé›¨å¤¹é›ª";  // Shower(s) of rain and snow mixed, slight
    case 84: return "ğŸŒ¨ï¸ ä¸­åˆ°å¤§é˜µé›¨å¤¹é›ª";  // Shower(s) of rain and snow mixed, moderate or heavy
    case 85: return "â„ï¸ å°é˜µé›ª";  // Snow shower(s), slight
    case 86: return "â„ï¸ ä¸­åˆ°å¤§é˜µé›ª";  // Snow shower(s), moderate or heavy
    case 87: return "â„ï¸ å°é˜µé›ªç²’æˆ–å°å†°é›¹";  // Shower(s) of snow pellets or small hail, slight
    case 88: return "â„ï¸ ä¸­åˆ°å¤§é˜µé›ªç²’æˆ–å°å†°é›¹";  // Shower(s) of snow pellets or small hail, moderate or heavy
    case 89: return "âšª å°é˜µå†°é›¹";  // Shower(s) of hail, slight
    case 90: return "âšª ä¸­åˆ°å¤§é˜µå†°é›¹";  // Shower(s) of hail, moderate or heavy
    case 91: return "âš¡ğŸŒ§ï¸ å°é›¨ï¼Œå‰ä¸€å°æ—¶æœ‰é›·æš´";  // Slight rain, thunderstorm during preceding hour
    case 92: return "âš¡ğŸŒ§ï¸ ä¸­åˆ°å¤§é›¨ï¼Œå‰ä¸€å°æ—¶æœ‰é›·æš´";  // Moderate or heavy rain, thunderstorm during preceding hour
    case 93: return "âš¡ğŸŒ¨ï¸ å°é›ªæˆ–é›¨å¤¹é›ªï¼Œå‰ä¸€å°æ—¶æœ‰é›·æš´";  // Slight snow, or rain and snow mixed or hail, thunderstorm during preceding hour
    case 94: return "âš¡ğŸŒ¨ï¸ ä¸­åˆ°å¤§é›ªæˆ–é›¨å¤¹é›ªï¼Œå‰ä¸€å°æ—¶æœ‰é›·æš´";  // Moderate or heavy snow, or rain and snow mixed or hail, thunderstorm during preceding hour
    case 95: return "âš¡ é›·æš´ï¼Œæ— å†°é›¹";  // Thunderstorm, slight or moderate, without hail
    case 96: return "âš¡âšª é›·æš´ä¼´æœ‰å†°é›¹";  // Thunderstorm, slight or moderate, with hail
    case 97: return "âš¡ å¼ºé›·æš´ï¼Œæ— å†°é›¹";  // Thunderstorm, heavy, without hail
    case 98: return "âš¡ğŸŒªï¸ é›·æš´ä¼´æ²™å°˜æš´";  // Thunderstorm combined with duststorm or sandstorm
    case 99: return "âš¡âšª å¼ºé›·æš´ä¼´æœ‰å†°é›¹";  // Thunderstorm, heavy, with hail
    default: return "â“ æœªçŸ¥å¤©æ°”ä»£ç ";  // Unknown weather code
    }
}